spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/GqlDatastoreQuery.java

```diff
@@ -0,0 +1,202 @@
+import org.springframework.data.repository.query.Parameters;
+import org.springframework.data.repository.query.QueryMethod;
+import org.springframework.data.repository.query.QueryMethodEvaluationContextProvider;
+import org.springframework.data.repository.query.RepositoryQuery;
+import org.springframework.expression.spel.standard.SpelExpressionParser;
+import org.springframework.util.StringUtils;
+
+/**
+ * Query Method for GQL queries.
+ *
+ * @author Chengyuan Zhao
+ *
+ * @since 1.1
+ */
+public class GqlDatastoreQuery<T> implements RepositoryQuery {
+
+	private static final Map<Class<?>, Function<Builder, BiFunction<String, Object, Builder>>>
+			GQL_PARAM_BINDING_FUNC_MAP;
+
+	static {
+		GQL_PARAM_BINDING_FUNC_MAP = ImmutableMap
+				.<Class<?>, Function<Builder, BiFunction<String, Object, Builder>>>builder()
+				.put(Cursor.class, builder -> (s, o) -> builder.setBinding(s, (Cursor) o))
+				.put(String.class, builder -> (s, o) -> builder.setBinding(s, (String) o))
+				.put(String[].class, builder -> (s, o) -> builder.setBinding(s, (String[]) o))
+				.put(Long.class, builder -> (s, o) -> builder.setBinding(s, (Long) o))
+				.put(long[].class, builder -> (s, o) -> builder.setBinding(s, (long[]) o))
+				.put(Double.class, builder -> (s, o) -> builder.setBinding(s, (Double) o))
+				.put(double[].class, builder -> (s, o) -> builder.setBinding(s, (double[]) o))
+				.put(Boolean.class,
+						builder -> (s, o) -> builder.setBinding(s, (Boolean) o))
+				.put(boolean[].class,
+						builder -> (s, o) -> builder.setBinding(s, (boolean[]) o))
+				.put(Timestamp.class,
+						builder -> (s, o) -> builder.setBinding(s, (Timestamp) o))
+				.put(Timestamp[].class,
+						builder -> (s, o) -> builder.setBinding(s, (Timestamp[]) o))
+				.put(Key.class, builder -> (s, o) -> builder.setBinding(s, (Key) o))
+				.put(Key[].class, builder -> (s, o) -> builder.setBinding(s, (Key[]) o))
+				.put(Blob.class, builder -> (s, o) -> builder.setBinding(s, (Blob) o))
+				.put(Blob[].class, builder -> (s, o) -> builder.setBinding(s, (Blob[]) o))
+				.build();
+	}
+
+	private final QueryMethod queryMethod;
+
+	private final DatastoreOperations datastoreOperations;
+
+	protected final DatastoreMappingContext datastoreMappingContext;
+
+	private final Class<T> entityType;
+
+	private final String gql;
+
+	// unused currently, but will be used for SpEL expression in the query.
+	private QueryMethodEvaluationContextProvider evaluationContextProvider;
+
+	// unused currently, but will be used for SpEL expression in the query.
+	private SpelExpressionParser expressionParser;
+
+	/**
+	 * Constructor
+	 * @param type the underlying entity type
+	 * @param queryMethod the underlying query method to support.
+	 * @param datastoreOperations used for executing queries.
+	 * @param datastoreMappingContext used for getting metadata about entities.
+	 */
+	GqlDatastoreQuery(Class<T> type, QueryMethod queryMethod,
```

Comment:
public/private?

Reply:
It can't be private because it's used elsewhere in the package. It doesn't need to be public because it's only used elsewhere in the package. package-private is appropriate.